packages/typescript/src/lib/server.ts

```diff
@@ -0,0 +1,80 @@
+import http from 'http';
+import fs from 'fs';
+import path from 'path';
+import { TypeServeOptions } from '../types';
+import { LoggerInstance } from '../Logger';
+
+export type TypeServerOptions = {
+  outputPath: string;
+  port: TypeServeOptions['port'];
+  host: TypeServeOptions['host'];
+  logger: LoggerInstance;
+};
+
+const activeServers: Map<number, http.Server | null> = new Map();
+
+export const startServer = async ({
+  outputPath,
+  port,
+  host,
+  logger,
+}: TypeServerOptions) => {
+  return new Promise((resolve) => {
+    if (activeServers.get(port!)) {
+      resolve(1);
+      return;
+    }
+
+    const server = http.createServer((req, res) => {
+      const { url: fileName } = req;
+
+      if (!fileName) {
+        logger.log(`Unable to find file: ${fileName} in request url`);
+        res.end();
+        return;
+      }
+
+      const safeSuffix = path
+        .normalize(fileName)
+        .replace(/^(\.\.(\/|\\|$))+/, '');
+      const safeJoin = path.join(outputPath, safeSuffix);
+
+      logger.log(`Retrieving file: ${safeJoin}`);
+
+      const fileStream = fs.createReadStream(safeJoin);
+
+      fileStream.on('error', (err) => {
+        logger.log(`Error reading file: ${err}`);
+        res.statusCode = 500;
+        res.end();
+      });
+
+      res.writeHead(200, { 'Content-Type': 'text/plain' });
+      fileStream.pipe(res);
```

Comment:
I found out some vulnerabilities here like:
- Directory Traversal Attacks: code is using path.normalize and path.join to build the file path, but it can be easily hacked if you put some more back steps to the relative path, e.g: `http://example.com:8080/../../../../../etc/passwd`.
- Error Handling: if something unexpected occurs during the file serving, or even trying to find it, the request will continue open, we need to use some defensive tactics here.

Here is a simple suggestion on how to solve it

```suggestion
      const fileName = path.join(outputPath, req.url);

      try {
        // Ensure the requested file is within the specified directory
        if (!fileName.startsWith(outputPath)) {
          res.writeHead(403, { 'Content-Type': 'text/plain' });
          res.end('Forbidden');
          return;
        }

        // Check if the file exists
        fs.stat(fileName, (err, stat) => {
          if (err) {
            logger.log(`Error reading file: ${err}`);
            res.writeHead(500, { 'Content-Type': 'text/plain' });
            res.end('Internal Server Error');
          } else {
            if (stat.isFile()) {
              res.writeHead(200, { 'Content-Type': 'text/plain' });
              fs.createReadStream(fileName).pipe(res);
            } else {
              // Handle non-file requests (e.g., directories)
              res.writeHead(404, { 'Content-Type': 'text/plain' });
              res.end('Not Found');
            }
          }
        });
      } catch (err) {
        logger.log(`Error reading file: ${err}`);
        res.writeHead(500, { 'Content-Type': 'text/plain' });
        res.end('Internal Server Error');
      }
```

Reply:
If we are looking to use it in production, it will need to have some more options, but if its only for dev envs, I believe the error handling suggestion is fine by now :smile: